File: frontend/scripts/compare-webgpu-wgpu.mjs
// CPU reference comparator for the wgpu-native WGSL parity run.
//
// Reads the RGBA frames produced by scripts/compare-webgpu-wgpu.py
// (/tmp/wgpu-parity/<case>.rgba) and diffs them against the C++-verified CPU
// fp64 core (local-render-core.renderLocalRgba). Run order:
//
//   /tmp/wgpu-venv/bin/python scripts/compare-webgpu-wgpu.py
//   node scripts/compare-webgpu-wgpu.mjs

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { spawnSync } from "node:child_process";
import ts from "typescript";

const ROOT = resolve(import.meta.dirname, "..");
const OUT_DIR = "/tmp/wgpu-parity";
const REFERENCE = process.env.REFERENCE_BINARY ?? resolve(ROOT, "../backend/build/browser_orbit_reference");

function dataUrl(relative) {
  const source = readFileSync(resolve(ROOT, relative), "utf8");
  const compiled = ts.transpileModule(source, {
    compilerOptions: { module: ts.ModuleKind.ESNext, target: ts.ScriptTarget.ES2022 },
  }).outputText;
  return `data:text/javascript;base64,${Buffer.from(compiled).toString("base64")}`;
}

const orbitUrl = dataUrl("src/lib/fractal/local-orbit-program.ts");
const cacheUrl = dataUrl("src/lib/fractal/local-field-cache.ts");
const coreSource = readFileSync(resolve(ROOT, "src/lib/fractal/local-render-core.ts"), "utf8");
const coreCompiled = ts.transpileModule(coreSource, {
  compilerOptions: { module: ts.ModuleKind.ESNext, target: ts.ScriptTarget.ES2022 },
}).outputText
  .replace('"./local-orbit-program"', JSON.stringify(orbitUrl))
  .replace('"./local-field-cache"', JSON.stringify(cacheUrl));
const core = await import(`data:text/javascript;base64,${Buffer.from(coreCompiled).toString("base64")}`);

const CASES = [
  { name: "mandel_escape", width: 128, height: 96, spec: { centerRe: -0.75, centerIm: 0, scale: 3, iterations: 180, variant: "mandelbrot", metric: "escape", colorMap: "classic_cos", colorMode: "direct", cyclesPerOctave: 1, smooth: false, rotationDeg: 0, julia: false, juliaRe: 0, juliaIm: 0, bailout: 2, pairwiseCap: 64 } },
  { name: "ship_julia_smooth", width: 128, height: 96, spec: { centerRe: -0.64, centerIm: 0.03, scale: 2.4, iterations: 180, variant: "burning_ship", metric: "escape", colorMap: "viridis", colorMode: "direct", cyclesPerOctave: 1, smooth: true, rotationDeg: 19, julia: true, juliaRe: -0.8, juliaIm: 0.156, bailout: 2, pairwiseCap: 64 } },
  { name: "min_abs_hs_rainbow", width: 128, height: 96, spec: { centerRe: -0.75, centerIm: 0, scale: 3, iterations: 200, variant: "mandelbrot", metric: "min_abs", colorMap: "hs_rainbow", colorMode: "direct", cyclesPerOctave: 1, smooth: false, rotationDeg: 0, julia: false, juliaRe: 0, juliaIm: 0, bailout: 2, pairwiseCap: 64 } },
  { name: "max_abs_twilight_smooth", width: 128, height: 96, spec: { centerRe: -0.75, centerIm: 0, scale: 3, iterations: 200, variant: "tricorn", metric: "max_abs", colorMap: "twilight", colorMode: "direct", cyclesPerOctave: 1, smooth: true, rotationDeg: 0, julia: false, juliaRe: 0, juliaIm: 0, bailout: 2, pairwiseCap: 64 } },
  { name: "envelope_inferno_rotated", width: 128, height: 96, spec: { centerRe: -0.5, centerIm: 0, scale: 2.2, iterations: 220, variant: "heart", metric: "envelope", colorMap: "inferno", colorMode: "direct", cyclesPerOctave: 1, smooth: false, rotationDeg: 37, julia: false, juliaRe: 0, juliaIm: 0, bailout: 2, pairwiseCap: 64 } },
  { name: "perp_ship_spectral_julia", width: 128, height: 96, spec: { centerRe: -0.6, centerIm: 0.1, scale: 2, iterations: 220, variant: "perp_ship", metric: "escape", colorMap: "spectral1530", colorMode: "direct", cyclesPerOctave: 1, smooth: true, rotationDeg: -90, julia: true, juliaRe: -0.8, juliaIm: 0.156, bailout: 2, pairwiseCap: 64 } },
  { name: "deep_zoom_fp32_stress", width: 128, height: 96, spec: { centerRe: -0.7435, centerIm: 0.1314, scale: 0.0002, iterations: 600, variant: "mandelbrot", metric: "escape", colorMap: "classic_cos", colorMode: "direct", cyclesPerOctave: 1, smooth: false, rotationDeg: 0, julia: false, juliaRe: 0, juliaIm: 0, bailout: 2, pairwiseCap: 64 } },
];

const DIAG_MAX_CHANGED_FRACTION = 0.25;
const MAX_DELTA = 255;

// Same-precision stage: the WGSL fp32 shader vs the C++ fp32 reference.
// Operation order mirrors the shader, so this must be near byte-exact.
// Independent fp32 implementations (naga/SPIR-V on the GPU vs g++ -O3 with
// FMA contraction) can still flip a handful of iteration-boundary pixels
// (measured 0.1%), so allow a small changed fraction; anything structural
// (e.g. a wrong orbit formula or palette) shows up as a large fraction.
const FP32_FRAMES = [
  { name: "mandel_escape", frame: "mandelbrot" },
  { name: "ship_julia_smooth", frame: "burning_ship_julia" },
  { name: "deep_zoom_fp32_stress", frame: "deep_zoom" },
];
const FP32_MAX_CHANGED_FRACTION = 0.005;

let failed = 0;
for (const entry of FP32_FRAMES) {
  const run = spawnSync(REFERENCE, ["--frame-fp32", entry.frame]);
  if (!run.stdout?.length) throw run.error ?? new Error(`native fp32 frame ${entry.frame} failed`);
  const gpu = readFileSync(`${OUT_DIR}/${entry.name}.rgba`);
  let maxDelta = 0;
  let changed = 0;
  for (let index = 0; index < run.stdout.length; index += 1) {
    const delta = Math.abs(gpu[index] - run.stdout[index]);
    maxDelta = Math.max(maxDelta, delta);
    if (delta > 1) changed += 1;
  }
  const changedFraction = changed / (gpu.length / 4);
  const ok = maxDelta <= 255 && changedFraction <= FP32_MAX_CHANGED_FRACTION;
  if (!ok) {
    console.error(`webgpu/native parity: ${entry.name} [fp32 ref] FAIL maxDelta=${maxDelta} changed=${changed}/${gpu.length / 4} (${(changedFraction * 100).toFixed(2)}%)`);
    failed += 1;
  } else {
    console.log(`webgpu/native parity: ${entry.name} [fp32 ref] passed (maxDelta=${maxDelta}, changed=${(changedFraction * 100).toFixed(2)}%)`);
  }
}

// Diagnostic stage: WGSL fp32 vs the C++-verified CPU fp64 core. fp32/fp64
// legitimately disagree on iteration-boundary pixels (all measured
// differences sit on field boundaries), so this stage only sanity-caps the
// changed fraction to catch structural coloring errors.

for (const entry of CASES) {
  const gpu = readFileSync(`${OUT_DIR}/${entry.name}.rgba`);
  if (gpu.length !== entry.width * entry.height * 4) {
    console.error(`${entry.name}: gpu frame size mismatch (${gpu.length})`);
    failed += 1;
    continue;
  }
  const cpu = core.renderLocalRgba(entry.spec, entry.width, entry.height);
  let maxDelta = 0;
  let changed = 0;
  let mean = 0;
  for (let index = 0; index < cpu.length; index += 1) {
    const delta = Math.abs(gpu[index] - cpu[index]);
    maxDelta = Math.max(maxDelta, delta);
    if (delta > 1) changed += 1;
    mean += delta;
  }
  const changedFraction = changed / entry.width / entry.height;
  const ok = maxDelta <= MAX_DELTA && changedFraction <= DIAG_MAX_CHANGED_FRACTION;
  if (!ok) {
    console.error(`webgpu/native parity: ${entry.name} [fp64 diag] FAIL maxDelta=${maxDelta} changed=${changed}/${entry.width * entry.height} (${(changedFraction * 100).toFixed(2)}%) mean=${(mean / cpu.length).toFixed(3)}`);
    failed += 1;
  } else {
    console.log(`webgpu/native parity: ${entry.name} [fp64 diag] ok (maxDelta=${maxDelta}, changed=${(changedFraction * 100).toFixed(2)}%, mean=${(mean / cpu.length).toFixed(3)})`);
  }
}
if (failed) {
  console.error(`webgpu/native parity: ${failed}/${CASES.length} cases failed`);
  process.exit(1);
}
console.log(`webgpu/native parity: ${CASES.length} cases passed (wgpu-native WGSL fp32 vs C++-verified fp64 core)`);

File: frontend/scripts/compare-webgpu-wgpu.py
#!/usr/bin/env python3
"""WebGPU WGSL parity runner (Python wgpu).

Extracts the exact FRACTAL_SHADER WGSL source from
frontend/src/lib/fractal/webgpu-renderer.ts, runs it through wgpu-native
(Vulkan, typically the NVIDIA adapter) for the same 7 cases as
compare-webgpu-render.mjs, and writes the raw RGBA output to
/tmp/wgpu-parity/<case>.rgba for the Node comparator to diff against the
C++-verified CPU fp64 core.

Usage:
  /tmp/wgpu-venv/bin/python scripts/compare-webgpu-wgpu.py
"""

import json
import math
import os
import re
import struct
import sys
from pathlib import Path

import wgpu

ROOT = Path(__file__).resolve().parent.parent
RENDERER = ROOT / "src" / "lib" / "fractal" / "webgpu-renderer.ts"
OUT_DIR = Path("/tmp/wgpu-parity")

# ---------------------------------------------------------------------------
# Extract the WGSL shader source verbatim from the TypeScript module.
# ---------------------------------------------------------------------------

source = RENDERER.read_text(encoding="utf-8")
match = re.search(r"FRACTAL_SHADER\s*=\s*/\*\s*wgsl\s*\*/\s*`(.*?)`", source, re.S)
if not match:
    sys.exit("FRACTAL_SHADER not found in " + str(RENDERER))
WGSL = match.group(1)

# ---------------------------------------------------------------------------
# Index tables mirroring webgpu-renderer.ts.
# ---------------------------------------------------------------------------

VARIANT_INDEX = {
    "mandelbrot": 0, "tricorn": 1, "burning_ship": 2, "celtic": 3, "heart": 4,
    "buffalo": 5, "perp_buffalo": 6, "celtic_ship": 7, "mandelceltic": 8, "perp_ship": 9,
}
METRIC_INDEX = {"escape": 0, "min_abs": 1, "max_abs": 2, "envelope": 3}
PALETTE_INDEX = {
    "classic_cos": 0, "mod17": 1, "hsv_wheel": 2, "tri765": 3, "grayscale": 4,
    "hs_rainbow": 5, "inferno": 6, "viridis": 7, "twilight": 8, "ember_blue": 9,
    "spectral1530": 10,
}

CASES = [
    ("mandel_escape", 128, 96, {"centerRe": -0.75, "centerIm": 0, "scale": 3, "iterations": 180, "variant": "mandelbrot", "metric": "escape", "colorMap": "classic_cos", "smooth": False, "rotationDeg": 0, "julia": False, "juliaRe": 0, "juliaIm": 0, "bailout": 2}),
    ("ship_julia_smooth", 128, 96, {"centerRe": -0.64, "centerIm": 0.03, "scale": 2.4, "iterations": 180, "variant": "burning_ship", "metric": "escape", "colorMap": "viridis", "smooth": True, "rotationDeg": 19, "julia": True, "juliaRe": -0.8, "juliaIm": 0.156, "bailout": 2}),
    ("min_abs_hs_rainbow", 128, 96, {"centerRe": -0.75, "centerIm": 0, "scale": 3, "iterations": 200, "variant": "mandelbrot", "metric": "min_abs", "colorMap": "hs_rainbow", "smooth": False, "rotationDeg": 0, "julia": False, "juliaRe": 0, "juliaIm": 0, "bailout": 2}),
    ("max_abs_twilight_smooth", 128, 96, {"centerRe": -0.75, "centerIm": 0, "scale": 3, "iterations": 200, "variant": "tricorn", "metric": "max_abs", "colorMap": "twilight", "smooth": True, "rotationDeg": 0, "julia": False, "juliaRe": 0, "juliaIm": 0, "bailout": 2}),
    ("envelope_inferno_rotated", 128, 96, {"centerRe": -0.5, "centerIm": 0, "scale": 2.2, "iterations": 220, "variant": "heart", "metric": "envelope", "colorMap": "inferno", "smooth": False, "rotationDeg": 37, "julia": False, "juliaRe": 0, "juliaIm": 0, "bailout": 2}),
    ("perp_ship_spectral_julia", 128, 96, {"centerRe": -0.6, "centerIm": 0.1, "scale": 2, "iterations": 220, "variant": "perp_ship", "metric": "escape", "colorMap": "spectral1530", "smooth": True, "rotationDeg": -90, "julia": True, "juliaRe": -0.8, "juliaIm": 0.156, "bailout": 2}),
    ("deep_zoom_fp32_stress", 128, 96, {"centerRe": -0.7435, "centerIm": 0.1314, "scale": 0.0002, "iterations": 600, "variant": "mandelbrot", "metric": "escape", "colorMap": "classic_cos", "smooth": False, "rotationDeg": 0, "julia": False, "juliaRe": 0, "juliaIm": 0, "bailout": 2}),
]


def uniform_bytes(spec: dict, width: int, height: int) -> bytes:
    angle = spec["rotationDeg"] * math.pi / 180
    floats = [
        spec["centerRe"], spec["centerIm"], spec["scale"], width / height,
        math.cos(angle), math.sin(angle), spec["bailout"], spec["bailout"] ** 2,
        spec["juliaRe"], spec["juliaIm"], 0.0, 0.0,
    ]
    u32s = [
        width, height, spec["iterations"],
        VARIANT_INDEX[spec["variant"]], METRIC_INDEX[spec["metric"]],
        PALETTE_INDEX[spec["colorMap"]], 1 if spec["smooth"] else 0,
        1 if spec["julia"] else 0,
    ]
    return struct.pack("<12f", *floats) + struct.pack("<8I", *u32s)


def main() -> int:
    adapter = wgpu.gpu.request_adapter_sync(power_preference="high-performance")
    if adapter is None:
        print("no wgpu adapter (Vulkan unavailable?)")
        return 1
    info = dict(adapter.info)
    print(f"adapter: {info.get('device', 'unknown')} ({info.get('adapter_type', '?')} / {info.get('backend_type', '?')})")

    device = adapter.request_device_sync()

    shader = device.create_shader_module(code=WGSL)
    pipeline = device.create_compute_pipeline(
        layout="auto",
        compute={"module": shader, "entry_point": "main"},
    )

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    for name, width, height, spec in CASES:
        params = device.create_buffer(
            size=80,
            usage=wgpu.BufferUsage.UNIFORM | wgpu.BufferUsage.COPY_DST,
        )
        pixels = device.create_buffer(
            size=width * height * 4,
            usage=wgpu.BufferUsage.STORAGE | wgpu.BufferUsage.COPY_SRC,
        )
        readback = device.create_buffer(
            size=width * height * 4,
            usage=wgpu.BufferUsage.COPY_DST | wgpu.BufferUsage.MAP_READ,
        )
        device.queue.write_buffer(params, 0, uniform_bytes(spec, width, height))

        bind_group = device.create_bind_group(
            layout=pipeline.get_bind_group_layout(0),
            entries=[
                {"binding": 0, "resource": params},
                {"binding": 1, "resource": pixels},
            ],
        )
        encoder = device.create_command_encoder()
        pass_ = encoder.begin_compute_pass()
        pass_.set_pipeline(pipeline)
        pass_.set_bind_group(0, bind_group)
        pass_.dispatch_workgroups(math.ceil(width / 8), math.ceil(height / 8))
        pass_.end()
        encoder.copy_buffer_to_buffer(pixels, 0, readback, 0, width * height * 4)
        device.queue.submit([encoder.finish()])

        readback.map_sync(wgpu.MapMode.READ)
        data = bytes(readback.read_mapped())
        readback.unmap()

        # u32 packed (r | g<<8 | b<<16) -> RGBA bytes, alpha 255, like the TS renderer.
        packed = struct.unpack(f"<{width * height}I", data)
        rgba = bytearray(width * height * 4)
        for index, color in enumerate(packed):
            offset = index * 4
            rgba[offset] = color & 0xFF
            rgba[offset + 1] = (color >> 8) & 0xFF
            rgba[offset + 2] = (color >> 16) & 0xFF
            rgba[offset + 3] = 255
        target = OUT_DIR / f"{name}.rgba"
        target.write_bytes(bytes(rgba))
        print(f"wrote {target} ({width * height} px)")

    return 0


if __name__ == "__main__":
    sys.exit(main())
